File: backend/internal/handler/brand.go
package handler

import (
	"net/http"

	"github.com/vibestore/backend/internal/dto"
	"github.com/vibestore/backend/internal/middleware"
	"github.com/vibestore/backend/internal/model"
	"github.com/vibestore/backend/internal/repository"
)

type BrandHandler struct {
	brandRepo *repository.BrandRepository
}

func NewBrandHandler(brandRepo *repository.BrandRepository) *BrandHandler {
	return &BrandHandler{brandRepo: brandRepo}
}

// GET /api/v1/store/brands
func (h *BrandHandler) ListPublic(w http.ResponseWriter, r *http.Request) {
	storeID := middleware.GetStoreID(r.Context())
	brands, err := h.brandRepo.FindActiveByStoreID(r.Context(), storeID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "INTERNAL_ERROR", "Failed to list brands")
		return
	}
	if brands == nil {
		brands = []model.Brand{}
	}
	writeJSON(w, http.StatusOK, brands)
}

// GET /api/v1/admin/brands
func (h *BrandHandler) List(w http.ResponseWriter, r *http.Request) {
	storeID := middleware.GetAdminStoreID(r.Context())
	brands, err := h.brandRepo.FindByStoreID(r.Context(), storeID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "INTERNAL_ERROR", "Failed to list brands")
		return
	}
	if brands == nil {
		brands = []model.Brand{}
	}
	writeJSON(w, http.StatusOK, brands)
}

// GET /api/v1/admin/brands/{id}
func (h *BrandHandler) Get(w http.ResponseWriter, r *http.Request) {
	storeID := middleware.GetAdminStoreID(r.Context())
	id, err := urlParamUint64(r, "id")
	if err != nil {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Invalid brand ID")
		return
	}
	brand, err := h.brandRepo.FindByID(r.Context(), id, storeID)
	if err != nil || brand == nil {
		writeError(w, http.StatusNotFound, "NOT_FOUND", "Brand not found")
		return
	}
	writeJSON(w, http.StatusOK, brand)
}

// POST /api/v1/admin/brands
func (h *BrandHandler) Create(w http.ResponseWriter, r *http.Request) {
	storeID := middleware.GetAdminStoreID(r.Context())

	var req dto.BrandRequest
	if err := decodeJSON(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Invalid request body")
		return
	}
	if req.Name == "" || req.Slug == "" {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Name and slug are required")
		return
	}

	brand := &model.Brand{
		StoreID:  storeID,
		Name:     req.Name,
		Slug:     req.Slug,
		LogoURL:  req.LogoURL,
		IsActive: req.IsActive,
	}

	if err := h.brandRepo.Create(r.Context(), brand); err != nil {
		writeError(w, http.StatusInternalServerError, "INTERNAL_ERROR", "Failed to create brand")
		return
	}

	writeJSON(w, http.StatusCreated, brand)
}

// PUT /api/v1/admin/brands/{id}
func (h *BrandHandler) Update(w http.ResponseWriter, r *http.Request) {
	storeID := middleware.GetAdminStoreID(r.Context())
	id, err := urlParamUint64(r, "id")
	if err != nil {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Invalid brand ID")
		return
	}

	existing, err := h.brandRepo.FindByID(r.Context(), id, storeID)
	if err != nil || existing == nil {
		writeError(w, http.StatusNotFound, "NOT_FOUND", "Brand not found")
		return
	}

	var req dto.BrandRequest
	if err := decodeJSON(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Invalid request body")
		return
	}

	existing.Name = req.Name
	existing.Slug = req.Slug
	existing.LogoURL = req.LogoURL
	existing.IsActive = req.IsActive

	if err := h.brandRepo.Update(r.Context(), existing); err != nil {
		writeError(w, http.StatusInternalServerError, "INTERNAL_ERROR", "Failed to update brand")
		return
	}

	writeJSON(w, http.StatusOK, existing)
}

// DELETE /api/v1/admin/brands/{id}
func (h *BrandHandler) Delete(w http.ResponseWriter, r *http.Request) {
	storeID := middleware.GetAdminStoreID(r.Context())
	id, err := urlParamUint64(r, "id")
	if err != nil {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Invalid brand ID")
		return
	}

	if err := h.brandRepo.Delete(r.Context(), id, storeID); err != nil {
		writeError(w, http.StatusInternalServerError, "INTERNAL_ERROR", "Failed to delete brand")
		return
	}

	writeJSON(w, http.StatusOK, map[string]string{"message": "Brand deleted"})
}

File: backend-bun/src/routes/brand.ts
import Elysia from "elysia";
import { tenantMiddleware } from "../middleware/tenant";
import { authMiddleware } from "../middleware/auth";
import * as brandRepo from "../repositories/brand";
import type { BrandRequest } from "../types";
import { toPublicBrand } from "../dto";

export const publicBrandRoutes = new Elysia({ prefix: "/api/v1" })
  .use(tenantMiddleware)
  .get("/store/brands", async ({ storeId }) => {
    const brands = await brandRepo.findActiveBrandsByStoreID(storeId);
    return brands.map(toPublicBrand);
  });

export const adminBrandRoutes = new Elysia({ prefix: "/api/v1/admin/brands" })
  .use(tenantMiddleware)
  .use(authMiddleware)
  .get("/", async ({ storeId }) => {
    return await brandRepo.findBrandsByStoreID(storeId);
  })
  .get("/:id", async ({ params, storeId, set }) => {
    const brand = await brandRepo.findBrandByID(Number(params.id), storeId);
    if (!brand) {
      set.status = 404;
      return { code: "NOT_FOUND", message: "Brand not found" };
    }
    return brand;
  })
  .post("/", async ({ body, storeId, set }) => {
    const req = body as BrandRequest;
    const id = await brandRepo.createBrand({ ...req, store_id: storeId });
    const brand = await brandRepo.findBrandByID(id, storeId);
    set.status = 201;
    return brand;
  })
  .put("/:id", async ({ params, body, storeId, set }) => {
    const existing = await brandRepo.findBrandByID(Number(params.id), storeId);
    if (!existing) {
      set.status = 404;
      return { code: "NOT_FOUND", message: "Brand not found" };
    }
    const req = body as BrandRequest;
    const updated = {
      ...existing,
      name: req.name,
      slug: req.slug,
      logo_url: req.logo_url ?? null,
      is_active: req.is_active,
    };
    await brandRepo.updateBrand(updated);
    return await brandRepo.findBrandByID(existing.id, storeId);
  })
  .delete("/:id", async ({ params, storeId, set }) => {
    const existing = await brandRepo.findBrandByID(Number(params.id), storeId);
    if (!existing) {
      set.status = 404;
      return { code: "NOT_FOUND", message: "Brand not found" };
    }
    await brandRepo.deleteBrand(existing.id, storeId);
    set.status = 204;
  });
